### PR TITLE
Fixing issue 4129: Unexpected error message when copying from csv files with mismatched num of columns

### PR DESCRIPTION
**Pull request by SterlingT3485**
# Description
This PR aims to solve issue [4129](https://github.com/kuzudb/kuzu/issues/4129#issuecomment-2331899406),

The solution is to make use of 
`bool SniffCSVNameAndTypeDriver::addRow(uint64_t, common::column_id_t columnCount)` inside `driver.cpp`, 

it dumbly returns true before, we could add a check of `columnCount` here, similar to what we did in 
`bool ParsingDriver::addRow(uint64_t /*rowNum*/, common::column_id_t columnCount)`. 

And also in `serial_csv_reader.cpp` the `numColumn` of `columnInfo` inside the `Reader` need to be initialized properly as `auto columnInfo = CSVColumnInfo(bindInput->expectedColumnNames.size() /* numColumns */, {} /* columnSkips */);`,before it was simply zero. 

A new field `"rowEmpty"` is introduced in `SniffCSVNameAndTypeDriver`, which is checked by `addRow` function to skip any empty row during the sniffing, similar to what we did in `ParsingDriver`.

After these changes, the mismatched num of columns problem can be detected when sniffing the header, as expected in the BUG description.

Three test cases are revised accordingly, which is `521/1219/1223`, they was expecting exception to be thrown in content before, now they should expect exception to be thrown in header.

Fixes # (4129)

# Contributor agreement

- [x] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

**Review by ray6080 — COMMENTED** [review]
Thanks! This generally looks good to me. Though we should do a refactoring on `SniffCSVNameAndTypeDriver`. The logic of `addRow` inside `SniffCSVNameAndTypeDriver` now duplicates the one in `ParsingDriver`.

Can u make `SniffCSVNameAndTypeDriver` extend `SerialParsingDriver`?
Let me know when u are done with the refactoring, I'd like to take another look at this PR.

**Comment by ray6080** [comment]
Btw, please make the PR title be more meaningful: 1) you don't need to include your username in the title; 2) issue ID is not the best reference as you don't know what it refers to unless you clicks into the issue.

**Comment by SterlingT3485** [comment]
Thank Guodong for the reviewing, refactored the `SniffCSVNameAndTypeDriver` to be extended from `SerialParsingDriver`

File: src/include/processor/operator/persistent/reader/csv/driver.h [hunk 1]
```diff
@@ -58,17 +60,20 @@ class SerialParsingDriver : public ParsingDriver {
     BaseCSVReader* getReader() override;
 };
 
-struct SniffCSVNameAndTypeDriver {
-    std::vector<std::pair<std::string, common::LogicalType>> columns;
-    std::vector<bool> sniffType;
-    // if the type isn't declared in the header, sniff it
+class SniffCSVNameAndTypeDriver : public SerialParsingDriver {
     SerialCSVReader* reader;
```
Comment (ray6080) [inline src/include/processor/operator/persistent/reader/csv/driver.h:64]: Isn't this shadowing the `reader` of `SerialParsingDriver` at L50? I don't think this is properly set during construction.

Comment (SterlingT3485) [reply src/include/processor/operator/persistent/reader/csv/driver.h:64]: let me try delete this` reader` thus the reader inherited from `SerialParsingDriver` is set properly

File: src/include/processor/operator/persistent/reader/csv/driver.h [hunk 2]
```diff
@@ -16,6 +16,8 @@ namespace processor {
 class BaseCSVReader;
 class ParsingDriver {
     common::DataChunk& chunk;
+
+protected:
     bool rowEmpty;
```
Comment (ray6080) [inline src/include/processor/operator/persistent/reader/csv/driver.h:21]: Can u move these two fields below functions? Our code convention is class member fields are below functions. For example https://github.com/kuzudb/kuzu/blob/master/src/include/storage/storage_manager.h.

Same for other driver class in this file.

**Review by ray6080 — APPROVED** [review]

File: src/include/processor/operator/persistent/reader/csv/driver.h (hunk 2 above)
Comment (SterlingT3485) [reply src/include/processor/operator/persistent/reader/csv/driver.h:21]: Thanks for pointing this out, let me fix it.

**Pull request merged**